src/k2/persistence/logStream/LogStream.cpp

```diff
@@ -0,0 +1,287 @@
+    });
+}
+
+
+seastar::future<> 
+LogStream::write(Payload payload){
+    return _write(std::move(payload), 1);
+}
+
+seastar::future<> 
+LogStream::_write(Payload payload, uint32_t writeToWAL){
+    if (!_logStreamCreate){
+        K2INFO("Log stream does not create");
+        throw std::runtime_error("Log stream does not create");
+    }
+
+    String sealed_plogId;
+    String new_plogId;
+    uint32_t sealed_offest = 0;
+    
+    // determine whether the current plog will excceed the size limit after the incoming writing operation
+    if (_plogInfo[writeToWAL].second + payload.getSize() >= PLOG_MAX_SIZE){
+        sealed_plogId = _plogInfo[writeToWAL].first;
+        sealed_offest = _plogInfo[writeToWAL].second;
+        // change to the next available plog id
+        if (writeToWAL == 0){
+            new_plogId = _metadataPlogPool[_metadataPlogPool.size()-1];
+            _metadataPlogPool.pop_back();
+        }
+        else{
+            new_plogId = _walPlogPool[_walPlogPool.size()-1];
+            _walPlogPool.pop_back();
+        }
+        _plogInfo[writeToWAL].first = new_plogId;
+        _plogInfo[writeToWAL].second = 0;
+    }
+
+    uint32_t current_offset = _plogInfo[writeToWAL].second;
+    _plogInfo[writeToWAL].second += payload.getSize();
+
+    // write the incoming payload to current plog
+    return _client.append(_plogInfo[writeToWAL].first, current_offset,  std::move(payload))
+    .then([&, writeToWAL, sealed_plogId, new_plogId, sealed_offest] (auto&& response){
+        auto& [status, resp] = response;
+        if (!status.is2xxOK()) {
+            K2INFO("unable to append the plog");
+            K2INFO(status);
+            throw std::runtime_error("unable to append the plog");
+        }
+        // If sealed_offest == 0, then it means after the incoming writing operation, the current plog is not excceed the size limit
+        if (sealed_offest == 0){
+            return seastar::make_ready_future<>();
+        }
+        // It the previous plog excceed the size limit, we will seal the previous plog and write the metadata to corrsponding place (Metadata Log Stream or CPO)
+        return _client.seal(sealed_plogId, sealed_offest)
+        .then([&, writeToWAL, new_plogId, sealed_offest] (auto&& response){
+            auto& [status, resp] = response;
+            if (!status.is2xxOK()) {
+                throw std::runtime_error("unable to seal a plog");
+            }
+
+            // Write the metadata to metadata log stream
+            if (writeToWAL){
+                Payload temp_payload([] { return Binary(4096); });
+                temp_payload.write(sealed_offest);
+                temp_payload.write(std::move(new_plogId));
+                return _write(std::move(temp_payload), 0);
+            }
+            // Write the metadata to CPO
+            else{
+                return _cpo.UpdateMetadataStreamLog(Deadline<>(_cpo_timeout()), _logStreamName, sealed_offest, std::move(new_plogId))

```

Comment:
So the CPO keeps track of every metadata plog? Is that what we want, or would it be better for the CPO to only store the root (first) metadata plog?

Reply:
Yes. The log stream will not communicate with the CPO frequently. If the Plog size = 1G, then we will update the CPO once if we have sealed 10M WAL Plogs